### PR TITLE
Support REPLACE INTO for unsharded tables

### PR DESCRIPTION
**Pull request by bbeaudreault**
I built REPLACE support by re-using all of the Insert constructs. Replace is basically a special cased insert, so I created a new field on the Insert struct, called Replace. Setting to true affects the way we rewrite the query (`replace into` instead of `insert into`), and ensures that we don't execute a replace against a sharded schema.

### CODE REVIEW

**Comment by bbeaudreault** [comment]
I added a test to `vtgatev3_test.py` for this, but I can't run that test suite locally right now. I'm working on that, but for now will respond to any issues that come up in Travis.

**Comment by bbeaudreault** [comment]
Looks like all tests are now passing in travis

File: data/test/vtgate/dml_cases.txt [hunk 1]
```diff
@@ -492,3 +492,148 @@
...
+      },
+      "Query": "select next :n values from seq",
+      "Value": [
+        null
+      ]
+    }
+  }
+}
+
+# insert unsharded, multi-val
```
Comment (sougou) [inline data/test/vtgate/dml_cases.txt]: replace unsharded. Similar changes needed below.

File: data/test/vtgate/dml_cases.txt [hunk 2]
```diff
@@ -492,3 +492,148 @@
...
+        1,
+        null
+      ]
+    }
+  }
+}
+
+# insert no column list
+"replace into user values(1, 2, 3)"
+"unsupported: REPLACE INTO with sharded schema"
```
Comment (sougou) [inline data/test/vtgate/dml_cases.txt]: Move unsupported cases to unsupported_cases.txt.

File: go/vt/sqlparser/ast.go [hunk 3]
```diff
@@ -251,21 +251,35 @@ func (node *Union) WalkSubtree(visit Visit) error {
...
+// Replaces are currently disallowed in sharded schemas because
+// of the implications the deletion part may have on vindexes.
 type Insert struct {
 	Comments Comments
 	Ignore   string
 	Table    *TableName
 	Columns  Columns
 	Rows     InsertRows
 	OnDup    OnDup
+	Replace  bool
```
Comment (sougou) [inline go/vt/sqlparser/ast.go]: For uniformity, I'd recommend using `Operation` or `Command` as a field, that's initialized to `"insert"` or `"replace"`. This way, you won't need a conditional at the time of printing. You'll see similar examples for this in operators, distinct, etc.

File: go/vt/sqlparser/sql.y [hunk 4]
```diff
@@ -229,6 +230,22 @@ insert_statement:
     $$ = &Insert{Comments: Comments($2), Ignore: $3, Table: $4, Columns: cols, Rows: Values{vals}, OnDup: OnDup($7)}
   }
 
+replace_statement:
+  REPLACE comment_opt into_table_name column_list_opt row_list
```
Comment (sougou) [inline go/vt/sqlparser/sql.y]: You can also change `INSERT` into `insert_or_replace` and make it return the correct string.

File: go/vt/vtgate/planbuilder/insert.go [hunk 5]
```diff
@@ -23,6 +23,9 @@ func buildInsertPlan(ins *sqlparser.Insert, vschema VSchema) (*engine.Route, err
 	if !table.Keyspace.Sharded {
 		return buildInsertUnshardedPlan(ins, table, vschema)
 	}
+	if ins.Replace {
```
Comment (sougou) [inline go/vt/vtgate/planbuilder/insert.go]: This would become `ins.Operation == sqlparser.ReplaceStr`.

**Comment by sougou** [comment]
Well. After I sent out the review, I found some problems. Basically, if the replace matched a unique key, it will end up deleting that row. However, vttablet cannot predict what that row is without running individual selects on every unique key the table has.

So, we'll have to either categorize replace as `PASS_DML`, or peform a select for each unique key and fetch the pk values so they can be added to the list of affected primary keys :(.

Here's a pathological example where replace deletes two other rows:
```
MariaDB [a]> create table a(id int, val1 int, val2 int, primary key(id), unique index v1(val1), unique index v2(val2));
Query OK, 0 rows affected (0.03 sec)

MariaDB [a]> insert into a values(1, 2, 3), (2, 3, 2);
Query OK, 2 rows affected (0.00 sec)
Records: 2  Duplicates: 0  Warnings: 0

MariaDB [a]> select * from a;
+----+------+------+
| id | val1 | val2 |
+----+------+------+
|  1 |    2 |    3 |
|  2 |    3 |    2 |
+----+------+------+
2 rows in set (0.00 sec)

MariaDB [a]> replace into a values(3, 2, 2);
Query OK, 3 rows affected (0.01 sec)

MariaDB [a]> select * from a;
+----+------+------+
| id | val1 | val2 |
+----+------+------+
|  3 |    2 |    2 |
+----+------+------+
1 row in set (0.00 sec)
```

In the above case, the stream comment needs to contain (1), (2), (3). Not just (1).

**Comment by bbeaudreault** [comment]
@sougou I pushed a commit to fix all your comments.  In terms of the problem you mentioned, I'm currently making progress towards the following changes:

- When parsing schemas in `table_info.fetchIndexes`, add a field on Index to denote whether an index is unique.
- within query_gen.go, create an GenerateInsertSubquery function like the ones for Delete and Update. This will formulate a select query to search for rows who have existing values of unique keys  that conflict with the values in the insert
- pass the resulting values to formulate the stream comment to ensure the stream comment covers not just the primary key, but also columns of unique keys

This seems pretty straightforward so far. Let me know if you have concerns.

**Comment by sougou** [comment]
Plan sounds good.

**Comment by bbeaudreault** [comment]
I'm going to close this for now. The REPLACE ended up being in some tooling, not a usual query people did. So it fell off my radar.  Will revive if it becomes an issue gain.

**Comment by sougou** [comment]
With RBR, we're thinking of allowing all DMLs. Just debating how to expose it.

**Comment by bbeaudreault** [comment]
@sougou you are actively working on v3 right now, and I just saw more usages of REPLACE crop up. Thoughts on adding support and the RBR passthrough?

**Pull request closed without merging**